gadgets/trace_sslsnoop/program.bpf.c

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)  TODO: Apache2?
+/* Copyright (c) 2024 The Inspektor Gadget authors */
+//
+// Based on sslsnoop from bpftrace by Tao Xu
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+#include <gadget/buffer.h>
+#include <gadget/macros.h>
+#include <gadget/mntns_filter.h>
+
+/* The maximum number of items in maps */
+#define MAX_ENTRIES 8192
+
+enum function {
+	SSL_read,
+	SSL_write,
+	SSL_do_handshake,
+
+	rsa_ossl_public_encrypt,
+	rsa_ossl_public_decrypt,
+	rsa_ossl_private_encrypt,
+	rsa_ossl_private_decrypt,
+	RSA_sign,
+	RSA_verify,
+	ossl_ecdsa_sign,
+	ossl_ecdsa_verify,
+	ossl_ecdh_compute_key,
+};
+
+struct event {
+	gadget_mntns_id mntns_id;
+	__u64 time;
+	__u32 pid;
+	__u32 tid;
+	__u8 comm[TASK_COMM_LEN];
+	enum function operation;
+	__u64 retval;
+	__u64 latency;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);   // tid
+	__type(value, u64); // start_time
+} start_ssl SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u32);   // tid
+	__type(value, u64); // start_time
+} start_crypto SEC(".maps");
+
+GADGET_TRACER_MAP(events, 1024 * 256);
+
+GADGET_TRACER(sslsnoop, events, event);
+
+static __always_inline int record_start_time(struct pt_regs *ctx, enum function operation)
+{
+	u64 mntns_id;
+	u64 start_time;
+	u32 tid;
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	start_time = bpf_ktime_get_ns();
+	tid = (__u32)bpf_get_current_pid_tgid();
+
+	switch (operation) {
+		case SSL_read:
+		case SSL_write:
+		case SSL_do_handshake:
+			bpf_map_update_elem(&start_ssl, &tid, &start_time, BPF_ANY);
+			break;
+		default:
+			bpf_map_update_elem(&start_crypto, &tid, &start_time, BPF_ANY);
+			break;
+	}
+	return 0;
+}
+
+static __always_inline int submit_event(struct pt_regs *ctx,
+					enum function operation, __u64 retval)
+{
+	u64 mntns_id;
+	struct event *event;
+	__u64 pid_tgid;
+	__u32 pid;
+	__u32 tid;
+	u64 current_time;
+	u64 start_time;
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	current_time = bpf_ktime_get_ns();
+
+	event = gadget_reserve_buf(&events, sizeof(*event));
+	if (!event)
+		return 0;
+
+	pid_tgid = bpf_get_current_pid_tgid();
+	pid = pid_tgid >> 32;
+	tid = (__u32)pid_tgid;
+
+	event->mntns_id = mntns_id;
+	event->time = current_time;
+	event->pid = pid;
+	event->tid = tid;
+	bpf_get_current_comm(event->comm, sizeof(event->comm));
+	event->operation = operation;
+	event->retval = retval;
+	switch (operation) {
+		case SSL_read:
+		case SSL_write:
+		case SSL_do_handshake: {
+			u64 *start_time_ptr = bpf_map_lookup_elem(&start_ssl, &tid);
+			if (!start_time_ptr)
+			{
+				gadget_discard_buf(event);
+				return 0;
+			}
+			start_time = *start_time_ptr;
+			bpf_map_delete_elem(&start_ssl, &tid);
```

Comment:
This could be done before `gadget_reserve_buf`, so if `gadget_reserve_buf` fails, we would not leak the entry in `start_ssl`. Also, that would remove the need for `gadget_discard_buf`, making the code simpler.

Ditto for `start_crypto` below.

If the process is killed between the "record_start" and the "submit_event", the maps would leak entries too. This could be prevented by checking in the `sched_process_exit` tracepoint for threads to clean up.

Reply:
> If the process is killed between the "record_start" and the "submit_event", the maps would leak entries too. This could be prevented by checking in the sched_process_exit tracepoint for threads to clean up.

Here we are using the `thread ID` instead of the `process ID`, because each thread may call a SSL function separately. Since the `sched_process_exit` is for process exiting, we can not use that here.

Reply:
You are right, it will be called when a thread exits.

Reply:
Credits to Falco for the sched_process_exit idea:
https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/programs/attached/events/sched_process_exit.bpf.c#L151

Reply:
This has been applied to all open pull requests.